internal/operator/bundle/install.go

```
@@ -0,0 +1,107 @@
+
+func (i *Install) BindFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&i.IndexImage, "index-image", defaultIndexImage, "index image in which to inject bundle")
+	fs.Var(&i.InstallMode, "install-mode", "install mode")
+	fs.StringVar(&i.InjectBundleMode, "mode", "", "mode to use for adding bundle to index")
+	_ = fs.MarkHidden("mode")
+}
+
+func (i Install) Run(ctx context.Context) (*v1alpha1.ClusterServiceVersion, error) {
+	if err := i.setup(ctx); err != nil {
+		return nil, err
+	}
+	return i.InstallOperator(ctx)
+}
+
+func (i *Install) setup(ctx context.Context) error {
+	labels, csv, err := loadBundle(ctx, i.BundleImage)
+	if err != nil {
+		return fmt.Errorf("load bundle: %v", err)
+	}
+
+	i.OperatorInstaller.PackageName = labels["operators.operatorframework.io.bundle.package.v1"]
+	i.OperatorInstaller.CatalogSourceName = fmt.Sprintf("%s-catalog", i.OperatorInstaller.PackageName)
+	i.OperatorInstaller.StartingCSV = csv.Name
+	i.OperatorInstaller.Channel = strings.Split(labels["operators.operatorframework.io.bundle.channels.v1"], ",")[0]
+
+	i.IndexImageCatalogCreator.InjectBundles = []string{i.BundleImage}
+	i.IndexImageCatalogCreator.InjectBundleMode = "replaces"
+	if i.IndexImageCatalogCreator.IndexImage == defaultIndexImage {
+		i.IndexImageCatalogCreator.InjectBundleMode = "semver"
+	}
+
+	return nil
+}
+
+func loadBundle(ctx context.Context, bundleImage string) (labels registryutil.Labels, csv *registry.ClusterServiceVersion, err error) {
+	bundlePath, err := registryutil.ExtractBundleImage(ctx, nil, bundleImage, false)
+	if err != nil {
+		return nil, nil, fmt.Errorf("pull bundle image: %v", err)
+	}
+	defer func() {
+		_ = os.RemoveAll(bundlePath)
+	}()
+
+	labels, _, err = registryutil.FindBundleMetadata(bundlePath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("load bundle metadata: %v", err)
+	}
+
+	relManifestsDir := labels["operators.operatorframework.io.bundle.manifests.v1"]
```

```suggestion
	relManifestsDir, _ := labels.GetManifestsDir()
```

---

@estroz WDYT of defaulting to `manifests` if the label is not found?

---

We should probably error out instead because the bundle is invalid.

---

Sounds good. I'll update to return an error.